docs/architecture/adr-011-generalize-genesis-accounts.md

```diff
@@ -0,0 +1,148 @@
+
+- 2019-08-30: initial draft
+
+## Context
+
+Summary: The `auth` module allows custom account types, but the `genaccounts` module does not.
+
+Currently the SDK allows for custom account types; the `auth` keeper stores any type fulfilling its Account interface. However `auth` does not handle exporting or loading accounts to/from a genesis file, this is done by `genaccounts`, which only handles one of 4 concrete account types (base, continuousVesting, delayedVesting, module).
+
+Projects desiring to use custom accounts (say custom vesting accounts) need to fork and modify `genaccounts`.
+
+
+## Decision
+
+In summary, we will (Un)Marshal all accounts (interface types) directly using amino, rather than converting to genaccount’s `GenesisAccount` type. Since doing this removes the majority of genaccount’s code, we will merge genaccounts into auth. Marshalled accounts will be stored in auth's genesis state.
+
+Detailed changes:
+
+#### 1) (un)marshal accounts directly using amino
+
+`auth.GenesisState` gains a new field `Accounts`
+```go
+// GenesisState - all auth state that must be provided at genesis
+type GenesisState struct {
+	Params   Params             `json:"params" yaml:"params"`
+	Accounts []exported.Account `json:"accounts" yaml:"accounts"`
+}
+```
+`auth.InitGenesis` and `auth.ExportGenesis` (un)marshal accounts as well as params.
+```go
+// InitGenesis - Init store state from genesis data
+func InitGenesis(ctx sdk.Context, ak AccountKeeper, data GenesisState) {
+	ak.SetParams(ctx, data.Params)
+	// load the accounts
+	for _, a := range data.Accounts {
+		acc := ak.NewAccount(ctx, a) // set account number
+		ak.SetAccount(ctx, acc)
+	}
+}
+
+// ExportGenesis returns a GenesisState for a given context and keeper
+func ExportGenesis(ctx sdk.Context, ak AccountKeeper) GenesisState {
+	params := ak.GetParams(ctx)
+	accounts := ak.GetAllAccounts(ctx)
+	return NewGenesisState(params, accounts)
+}
+```
+
+#### 2) registering custom account types on the auth codec
+
+The auth codec must have all custom account types registered to marshal them. We will follow the pattern established in gov for proposals.
+
+A custom account definition:
+```go
+import authTypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+// Register the module account type with the auth module codec so it can decode module accounts stored in a genesis file
+func init() {
+	authtypes.RegisterAccountTypeCodec(ModuleAccount{}, "cosmos-sdk/ModuleAccount")
+}
+type ModuleAccount struct {
+    ...
+```
+The auth codec definition:
+```go
+var ModuleCdc *codec.Codec
+func init() {
+    ModuleCdc = codec.New()
+    // register module msg's and Account interface
+    ...
+	// leave the codec unsealed
+}
+// RegisterAccountTypeCodec registers an external account type defined in another module for the internal ModuleCdc.
+func RegisterAccountTypeCodec(o interface{}, name string) {
+	ModuleCdc.RegisterConcrete(o, name, nil)
+}
+```
+
+#### 3) genesis validation for custom account types
```

Comment:
Strongly advise against this. Instead, we should create a `GenesisAccount`:

```go
type GenesisAccount interface {
  exported.Account
  Validate() error
}
```

Reply:
that's a nicer approach, I've updated the ADR.